### DISCUSSION TITLE
Handling wildcard SMILES in ChEBI classes

### DISCUSSION THREAD

**Discussion by rwst** [rdkit/rdkit#4359] [Q&A] [answered]
Rdkit parses the wildcard `*` as a single atom, see e.g. #3117. However, ChEBI apparently uses it to mean any moiety in their classes. Example:
https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:15889 (sterol)
SMILES: `C12C(C3C(C(CC3)*)(C)CC1)CCC4C2(CCC(C4)O)C`

How to match a molecule to this pattern, assuming `*` means any moiety? My idea would be to remove the wildcard, convert the pattern to SMARTS notation where every atom has explicit binding defined (using `D`, e.g. `[R0;D2]`), except for those binding the wildcard. Is something like this possible with rdkit, or even implemented?

**Comment by greglandrum** [comment, marked as answer]
As you've discovered, a `*` atom in a SMILES only matches another dummy atom (i.e. something with atomic number zero):
```
>>> q = Chem.MolFromSmiles('CC*')
>>> Chem.MolFromSmiles('CCC').HasSubstructMatch(q)
False
>>> Chem.MolToSmarts(q)
'[#6]-[#6]-[#0]'
```
The easiest way to get the behavior you want is to use `AdjustQueryProperties()`:
```
>>> q2 =Chem.AdjustQueryProperties(q)
>>> Chem.MolToSmarts(q2)
'[#6]-[#6]-*'
>>> Chem.MolFromSmiles('CCC').HasSubstructMatch(q2)
True
```

Note that the default behavior of `AdjustQueryProperties()` is also to add degree queries to ring atoms (preventing additional ring substitutions):
```
>>> Chem.MolToSmarts(q)
'[#6]1-[#6]-[#6]-[#6]-1-[#0]'
>>> Chem.MolToSmarts(Chem.AdjustQueryProperties(q))
'[#6&D2]1-[#6&D2]-[#6&D2]-[#6&D3]-1-*'
```
but you can disable this if you like:
```
>>> params = Chem.AdjustQueryParameters()
>>> params.adjustDegree = False
>>> Chem.MolToSmarts(Chem.AdjustQueryProperties(q,params))
'[#6]1-[#6]-[#6]-[#6]-1-*'
```

**Reply by rwst** [reply]
Looks good, thanks, but also catches sterol esters. Minimal example:
```
In [3]: q = Chem.MolFromSmiles('OC*')                                                                   

In [4]: q2 = Chem.AdjustQueryProperties(q)                                                              

In [5]: Chem.MolToSmarts(q)                                                                             
Out[5]: '[#8]-[#6]-[#0]'

In [6]: Chem.MolToSmarts(q2)                                                                            
Out[6]: '[#8]-[#6]-*'

In [7]: Chem.MolToSmarts(Chem.AdjustQueryProperties(q))                                                 
Out[7]: '[#8]-[#6]-*'

In [9]: Chem.MolFromSmiles('OCC').HasSubstructMatch(q2)                                                 
Out[9]: True

In [10]: Chem.MolFromSmiles('COCC').HasSubstructMatch(q2)                                               
Out[10]: True
```
q2 should have been converted to something like this:
```
In [16]: q3 = Chem.MolFromSmarts('[#8&D1]-[#6&D2]-*')                                                   

In [17]: Chem.MolFromSmiles('OCC').HasSubstructMatch(q3)                                                
Out[17]: True

In [18]: Chem.MolFromSmiles('COCC').HasSubstructMatch(q3)                                               
Out[18]: False
```
Report a bug?

**Reply by greglandrum** [reply]
It's not a bug, you just need to tell `AdjustQueryProperties()` which atoms to modify. As I mentioned in my first response, by default it adds the degree queries to ring atoms. You can change this though:
```
>>> q = Chem.MolFromSmiles('CC*')
>>> ps = Chem.AdjustQueryParameters()
>>> ps.adjustDegreeFlags = Chem.AdjustQueryWhichFlags.ADJUST_IGNOREDUMMIES
>>> q2 = Chem.AdjustQueryProperties(q,ps)
>>> Chem.MolToSmarts(q2)
'[#6&D1]-[#6&D2]-*'
```

**Reply by rwst** [reply]
Very nice, thank you. This finds 278 sterols in Wikidata (as defined by ChEBI).

**Reply by rwst** [reply]
How would/could this kind of matching be called? Is there a term for it?

**Reply by greglandrum** [reply]
I'm not sure what you mean. I would just call it substructure matching
The use of AdjustQueryProperties just ensures that the query is doing what you intend, since SMILES isn't expressive enough to express queries correctly.
